tests/assert_type/views/decorators/test_csrf.py

```diff
@@ -0,0 +1,46 @@
+from typing import Callable
+
+from django.http.request import HttpRequest
+from django.http.response import HttpResponse
+from django.views.decorators.csrf import csrf_protect
+from typing_extensions import assert_type
+
+
+@csrf_protect
+def good_view_positional(request: HttpRequest, /) -> HttpResponse:
+    return HttpResponse()
+
+
+# `assert_type` can only be used when `request` is pos. only.
+assert_type(good_view_positional, Callable[[HttpRequest], HttpResponse])
+
+
+# The decorator works too if `request` is not explicitly pos. only.
+@csrf_protect
+def good_view(request: HttpRequest) -> HttpResponse:
```

Comment:
Ignoring `AuthenticatedHttpRequest` for the moment, the added `test_csrf.py` is failing for me with mypy... Do you know what's up with that?

> tests/assert_type/views/decorators/test_csrf.py:19: error: Value of type variable "_ViewFuncT" of "csrf_protect" cannot be "Callable[[HttpRequest], HttpResponse]"  [type-var]
> tests/assert_type/views/decorators/test_csrf.py:24: error: Value of type variable "_ViewFuncT" of "csrf_protect" cannot be "Callable[[HttpRequest], Coroutine[Any, Any, HttpResponse]]"  [type-var]
> tests/assert_type/views/decorators/test_csrf.py:29: error: Value of type variable "_ViewFuncT" of "csrf_protect" cannot be "Callable[[HttpRequest, int, str], HttpResponse]"  [type-var]
> tests/assert_type/views/decorators/test_csrf.py:34: error: Value of type variable "_ViewFuncT" of "csrf_protect" cannot be "Callable[[HttpRequest, int, str], Coroutine[Any, Any, HttpResponse]]"  [type-var]

Reply:
I wanted to look into what could be done about `AuthenticatedHttpRequest`, but this is blocking that.

Reply:
Problem is probably the forced positional of `request`

Reply:
Indeed. Removing the `/,` from `_View` signature fixes that, but then makes line 9 fail.

We can fix this by duplicating both View protocols for both positional/non-positional cases. I think that would be fine. Unless there are better solutions?

```python
class _View(Protocol):
    def __call__(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponseBase: ...

class _AsyncView(Protocol):
    async def __call__(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponseBase: ...

class _ViewPositionalRequest(Protocol):
    def __call__(self, request: HttpRequest, /, *args: Any, **kwargs: Any) -> HttpResponseBase: ...

class _AsyncViewPositionalRequest(Protocol):
    async def __call__(self, request: HttpRequest, /, *args: Any, **kwargs: Any) -> HttpResponseBase: ...

_ViewFuncT = TypeVar(
    "_ViewFuncT", bound=_View | _ViewPositionalRequest | _AsyncView | _AsyncViewPositionalRequest
)  # noqa: PYI018
```

Reply:
After this change, the `AuthenticatedHttpRequest` hack works also, if I add this to `test_csrf.py` file:

```python
from django.contrib.auth.models import User


class AuthenticatedHttpRequest(HttpRequest):
    user: User


@csrf_protect  # type: ignore[type-var]  # pyright: ignore[reportArgumentType, reportUntypedFunctionDecorator]
def bad_view_with_arguments(request: AuthenticatedHttpRequest) -> HttpResponse:
    return HttpResponse()
```

Reply:
I don't know what is best to do here really.. Django runs all this with a first, forced positional, `HttpRequest` argument. But it's annoying that _everyone_ then _have_ to add the forced positional marker(`/`) to avoid violating the protocol.

An additional thing to keep in mind with the protocols here is that if the `request` argument isn't a forced positional it dictates the name of the argument to be "request".

i.e. trying something like below will yield an error:

```python
@csrf_protect
def someview(req: HttpRequest) -> HttpResponse: ...
```

---

An additional approach here could be to dig deeper into what attributes the decorator expects. For instance, I think, `@csrf_protect` "just" requires an object of something has the attributes `session` and `COOKIES` instead of the `HttpRequest` type. But it requires it as the first argument.

I'm not sure how much or if that can help, I think we still would have to decide on forced positional or not, but just wanted to mention it as an alternate approach to the `HttpRequest` type